### ISSUE TITLE
ProfilerAPI: Is it safe to cache IMetaDataImport

### ISSUE THREAD

**Issue by iskiselev** [dotnet/coreclr#25789]
Is it safe for profiler to cache IMetaDataImport for a module for a module lifetime?
I'm asking this question, as ICorProfilerInfo::GetModuleMetaData will fail if it is called from P/Invoke  callback, but looks like calls for cached IMetaDataImport methods works fine even from P/Invoke. 
MSDN documentation does not give any hints about IMetaDataImport lifetime and any problems that may be introduced by preserving IMetaDataImport non-released.

**Comment by tommcdon** [comment]
@davmason

**Comment by davmason** [comment]
Yes, that is safe as long as you respect COM semantics. Once the refcount goes to 0 it will no longer be usable, but as long as you keep a refcount on it you can use it safely.

What HR do you get when calling GetModuleMetadata? I wouldn't expect that to fail at any point when the module is alive. You may be hitting some corner case though of the runtime being in a weird state.

**Comment by mjsabby** [comment]
It's probably because `GetModuleMetadata` is `PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX`

**Comment by davmason** [comment]
Thanks @mjsabby, that's definitely it 😄
